pysmt/substituter.py

```diff
@@ -99,6 +98,80 @@ def _push_with_children_to_stack(self, formula, **kwargs):
                                                                  formula,
                                                                  **kwargs)
 
+    def substitute(self, formula, subs):
+        """Replaces any subformula in formula with the definition in subs."""
+
+        # Check that formula is a term
+        if not formula.is_term():
+            raise TypeError("substitute() can only be used on terms.")
+
+        for (i, k) in enumerate(subs):
+            v = subs[k]
+            # Check that substitutions are terms
+            if not k.is_term():
+                raise TypeError(
+                    "Only terms should be provided as substitutions." +
+                    " Non-term '%s' found." % k)
+            if not v.is_term():
+                raise TypeError(
+                    "Only terms should be provided as substitutions." +
+                    " Non-term '%s' found." % v)
+            # Check that substitutions belong to the current formula manager
+            if k not in self.manager:
+                raise TypeError(
+                    "Key %d does not belong to the Formula Manager." % i)
+            if v not in self.manager:
+                raise TypeError(
+                    "Value %d does not belong to the Formula Manager." % i)
+
+        self.orig_subs = subs
+        return self.walk(formula, substitutions=subs)
+
+
+class MGSubstituter(Substituter):
+    """Performs Most Specific Substitution.
+
+    This is the default behavior sinve version 0.5
+    """
+    def __init__(self, env):
+        Substituter.__init__(self, env=env)
+
+        self.set_function(self.walk_identity_or_replace, *op.all_types())
+        self._inner_idw = pysmt.walkers.IdentityDagWalker(env=self.env)
+
+
+    def walk_identity_or_replace(self, formula, args, **kwargs):
+        """If the formula appears in the substitution, return the substitution.
+           Otherwise, rebuild the formula, with its new children.
+        """
+        if formula in self.orig_subs:
+            return self.orig_subs[formula]
+        else:
+            return self._inner_idw.functions[formula.node_type()](formula, args=args)
```

Comment:
It took me a while to realize how this works, maybe a line of comment on the use of args....

Moreover, you should also pass '''**kwargs''' to the inner function.